channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <winpr/crt.h>
+#include <winpr/synch.h>
+#include <winpr/thread.h>
+#include <winpr/stream.h>
+
+#include <freerdp/channels/wtsvc.h>
+#include <freerdp/channels/log.h>
+
+#include "rdpgfx_common.h"
+#include "rdpgfx_main.h"
+
+#define TAG CHANNELS_TAG("rdpgfx.server")
+#define RDPGFX_RESET_GRAPHICS_PDU_SIZE 340
+
+/**
+ * Function description
+ * Create new stream for rdpgfx packet. The new stream length
+ * would be required data length + header. The header will be written
+ * to the stream before return, but the pduLength field might be 
+ * changed in rdpgfx_server_packet_send.
+ *
+ * @return new stream
+ */
+static wStream* rdpgfx_server_packet_new(UINT16 cmdId, UINT32 dataLen)
+{
+	UINT error;
+	wStream* s;
+	RDPGFX_HEADER header;
+
+	header.flags = 0;
+	header.cmdId = cmdId;
+	header.pduLength = RDPGFX_HEADER_SIZE + dataLen;
+
+	s = Stream_New(NULL, header.pduLength);
+
+	if(!s)
+	{
+	    WLog_ERR(TAG, "Stream_New failed!");
+	    return NULL;
+	}
+
+	/* Write header. Note that actual length will be filled
+	 * after the entire packet has been constructed. */
+	if ((error = rdpgfx_write_header(s, &header)))
+	{
+		WLog_ERR(TAG, "rdpgfx_write_header failed with error %lu!", error);
+		return NULL;
+	}
+
+	return s;
+}
+
+/**
+ * Function description
+ * Send the stream for rdpgfx packet. The packet would be compressed
+ * according to [MS-RDPEGFX].
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_server_packet_send(RdpgfxServerContext* context, wStream* s)
+{
+	UINT error;
+	UINT32 flags = 0;
+	ULONG written;
+	BYTE* pSrcData = Stream_Buffer(s);
+	UINT32 SrcSize = Stream_GetPosition(s);
+	wStream* fs;
+	
+	/* Fill actual length */
+	Stream_SetPosition(s, sizeof(RDPGFX_HEADER) - sizeof(UINT32));
+	Stream_Write_UINT32(s, SrcSize); /* pduLength (4 bytes) */
+
+	/* Allocate new stream with enough capacity. Additional overhead is
+	 * descriptor (1 bytes) + segmentCount (2 bytes) + uncompressedSize (4 bytes)
+	 * + segmentCount * size (4 bytes) */
+	fs = Stream_New(NULL, SrcSize + 7 
+	   + (SrcSize/ZGFX_SEGMENTED_MAXSIZE + 1) * 4);
+
+	if (zgfx_compress_to_stream(context->priv->zgfx, fs, pSrcData, SrcSize, &flags) < 0)
+	{
+		WLog_ERR(TAG, "zgfx_compress_to_stream failed!");
+		error = ERROR_INTERNAL_ERROR;
+		goto out;
+	}
+
+	error = WTSVirtualChannelWrite(context->priv->rdpgfx_channel, (PCHAR) Stream_Buffer(fs), 
+								   Stream_GetPosition(fs), &written) 
+								   ? CHANNEL_RC_OK : ERROR_INTERNAL_ERROR;
+
```

Comment:
OK to check written.
What do you mean 'not very clean' besides checking written?